baking content files doesn't work with UTF-8 encoding and changed metadata

I have a simple site with Freemarker templates and some simple HTML content files. 

When I bake the site, some content files are not written to the output directory. I played a bit with the content and the files and, from what I can figure out, it has something to do with encoding of the content files and the header metadata.

I can reproduce it on the `example_project_freemarker.zip` project like this:
- go to an empty folder
- run `jbake -i`
- go into the `content/about.html` file, change the file encoding to UTF-8 and the headers like this:
```
type=page
status=published
title=About
~~~~~~
```
- run `jbake -b`

The `about.html` will never be written in the output folder.

I'm using jbake v2.6.4, on Windows, with Java 1.8.

In the console I have some warnings:
```
19:08:38.875 WARN  org.jbake.template.TemplateEngines - unable to load engine
java.lang.reflect.InvocationTargetException: null
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)

        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstruct
orAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingC
onstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:408)
        at org.jbake.template.TemplateEngines.tryLoadEngine(TemplateEngines.java
:79)
        at org.jbake.template.TemplateEngines.registerEngine(TemplateEngines.jav
a:111)
        at org.jbake.template.TemplateEngines.loadEngines(TemplateEngines.java:1
02)
        at org.jbake.template.TemplateEngines.<init>(TemplateEngines.java:51)
        at org.jbake.template.DelegatingTemplateEngine.<init>(DelegatingTemplate
Engine.java:39)
        at org.jbake.app.Renderer.<init>(Renderer.java:82)
        at org.jbake.app.UtensilsFactory.createDefaultUtensils(UtensilsFactory.j
ava:26)
        at org.jbake.app.Oven.<init>(Oven.java:71)
        at org.jbake.launcher.Baker.bake(Baker.java:31)
        at org.jbake.launcher.Main.run(Main.java:97)
        at org.jbake.launcher.Main.run(Main.java:83)
        at org.jbake.launcher.Main.main(Main.java:56)
Caused by: java.lang.VerifyError: Bad <init> method call from inside of a branch

Exception Details:
  Location:
    org/codehaus/groovy/control/customizers/ASTTransformationCustomizer.<init>(L
java/lang/Class;)V @171: invokespecial
  Reason:
    Error exists in the bytecode
  Bytecode:
    0x0000000: b800 234d 05bd 006e 5903 2b53 5904 2c12
    0x0000010: c132 2bb9 0072 0200 5359 10ff 1202 b800
    0x0000020: 762a 5fab 0000 01af 0000 000a b951 be33
    0x0000030: 0000 0059 bd3c 11f8 0000 008e e4d2 7468
    0x0000040: 0000 00b7 f226 4f65 0000 00ca 1cb4 ddcd
    0x0000050: 0000 00df 224d 0e57 0000 00fe 4441 e4ca
    0x0000060: 0000 0127 4c34 1502 0000 0148 589a 5432
    0x0000070: 0000 0167 6648 e9fd 0000 0184 5f5a 5903
    0x0000080: 3212 7eb8 0051 c000 7e5f 5904 32b8 0042
    0x0000090: c000 445f 5905 32b8 0082 c000 785f 5906
    0x00000a0: 3212 84b8 0051 c000 845f 57b7 0087 a701
    0x00000b0: 2e5f 5a59 0332 b800 42c0 0044 5f59 0432
    0x00000c0: b800 82c0 0078 5f59 0532 1284 b800 51c0
    0x00000d0: 0084 5f57 b700 89a7 0105 5f5a 5903 32b8
    0x00000e0: 0042 c000 445f 57b7 008c a700 f25f 5a59
    0x00000f0: 0332 124b b800 51c0 004b 5f57 b700 8fa7
    0x0000100: 00dd 5f5a 5903 3212 7eb8 0051 c000 7e5f
    0x0000110: 5904 32b8 0042 c000 445f 57b7 0092 a700
    0x0000120: be5f 5a59 0332 127e b800 51c0 007e 5f59
    0x0000130: 0432 b800 42c0 0044 5f59 0532 b800 82c0
    0x0000140: 0078 5f57 b700 95a7 0095 5f5a 5903 3212
    0x0000150: 7eb8 0051 c000 7e5f 5904 3212 4bb8 0051
    0x0000160: c000 4b5f 57b7 0098 a700 745f 5a59 0332
    0x0000170: b800 42c0 0044 5f59 0432 1284 b800 51c0
    0x0000180: 0084 5f57 b700 9ba7 0055 5f5a 5903 32b8
    0x0000190: 0042 c000 445f 5904 32b8 0082 c000 785f
    0x00001a0: 57b7 009d a700 385f 5a59 0332 127e b800
    0x00001b0: 51c0 007e 5f59 0432 b800 42c0 0044 5f59
    0x00001c0: 0532 1284 b800 51c0 0084 5f57 b700 a0a7
    0x00001d0: 000d bb00 a259 12a4 b700 a7bf 57b1
  Stackmap Table:
    full_frame(@124,{UninitializedThis,Object[#68],Object[#122]},{Object[#124],U
ninitializedThis})
    full_frame(@177,{UninitializedThis,Object[#68],Object[#122]},{Object[#124],U
ninitializedThis})
    full_frame(@218,{UninitializedThis,Object[#68],Object[#122]},{Object[#124],U
ninitializedThis})
    full_frame(@237,{UninitializedThis,Object[#68],Object[#122]},{Object[#124],U
ninitializedThis})
    full_frame(@258,{UninitializedThis,Object[#68],Object[#122]},{Object[#124],U
ninitializedThis})
    full_frame(@289,{UninitializedThis,Object[#68],Object[#122]},{Object[#124],U
ninitializedThis})
    full_frame(@330,{UninitializedThis,Object[#68],Object[#122]},{Object[#124],U
ninitializedThis})
    full_frame(@363,{UninitializedThis,Object[#68],Object[#122]},{Object[#124],U
ninitializedThis})
    full_frame(@394,{UninitializedThis,Object[#68],Object[#122]},{Object[#124],U
ninitializedThis})
    full_frame(@423,{UninitializedThis,Object[#68],Object[#122]},{Object[#124],U
ninitializedThis})
    full_frame(@466,{UninitializedThis,Object[#68],Object[#122]},{Object[#124],U
ninitializedThis})
    full_frame(@476,{Object[#2],Object[#68],Object[#122]},{Object[#124]})

        at groovy.text.markup.MarkupTemplateEngine.<init>(MarkupTemplateEngine.j
ava:94)
        at groovy.text.markup.MarkupTemplateEngine.<init>(MarkupTemplateEngine.j
ava:125)
        at org.jbake.template.GroovyMarkupTemplateEngine.initializeTemplateEngin
e(GroovyMarkupTemplateEngine.java:53)
        at org.jbake.template.GroovyMarkupTemplateEngine.<init>(GroovyMarkupTemp
lateEngine.java:41)
        ... 16 common frames omitted
19:08:39.140 INFO  org.jbake.app.Oven - Baking has started...
```
and then this:

```
19:08:42.541 WARN  org.jbake.parser.MarkupEngine - Parsing skipped (missing type
 or status value in header meta data) for file D:\test\content\about.html!
19:08:42.541 WARN  org.jbake.app.Crawler - D:\test\content\about.html has an
 invalid header, it has been ignored!
```
**This last warning is false since the file has those header metadata. But maybe because `type` is first and UTF-8 has BOM characters somehow it doesn't register properly ?!?!?**

--- a/.travis.yml
+++ b/.travis.yml
@@ -37,9 +37,7 @@ jobs:
   - stage: Quality Check
     jdk: oraclejdk8
     script:
-    - "./gradlew jacocoRootReport"
-    after_success:
-    - "./gradlew sonarqube -i --stacktrace"
+    - "./gradlew jacocoRootReport sonarqube -i --stacktrace"
 
 notifications:
   irc: "irc.freenode.org#jbake"
@@ -61,3 +59,4 @@ cache:
      - "$HOME/.gradle/wrapper/"
      - "$HOME/.sonar/cache"
      - ".gradle"
+     - "build"

--- a/build.gradle
+++ b/build.gradle
@@ -39,6 +39,7 @@ ext {
     isTravis = (System.getenv("TRAVIS") == "true")
     isTravisPullRequest = (System.getenv("TRAVIS_PULL_REQUEST")) != "false"
     pullRequestId = System.getenv("TRAVIS_PULL_REQUEST")
+    pullRequestBranch = System.getenv("TRAVIS_PULL_REQUEST_BRANCH")
     hasGithub = System.getenv("GITHUBTOKEN") && System.getenv("GITHUBREPO")
     hasSonar = System.getenv("SONARORG") && System.getenv("SONARLOGIN")
     sonarDefaultURL = "https://sonarcloud.io"

--- a/gradle/sonarqube.gradle
+++ b/gradle/sonarqube.gradle
@@ -15,7 +15,7 @@ if ( JavaVersion.current().isJava8Compatible() ) {
             property "sonar.organization", System.getenv('SONARORG')
             property "sonar.login", System.getenv('SONARLOGIN')
             property "sonar.projectKey", sonarProjectKey
-
+            property "sonar.coverage.jacoco.xmlReportPaths", "${rootDir}/build/reports/jacoco/jacocoRootReport/jacocoRootReport.xml"
 
             if (isTravis && isTravisPullRequest) {
 

--- a/gradle/sonarqube.gradle
+++ b/gradle/sonarqube.gradle
@@ -10,7 +10,6 @@ if ( JavaVersion.current().isJava8Compatible() ) {
         }
 
         properties {
-
             property "sonar.host.url", sonarURL
             property "sonar.organization", System.getenv('SONARORG')
             property "sonar.login", System.getenv('SONARLOGIN')
@@ -18,17 +17,10 @@ if ( JavaVersion.current().isJava8Compatible() ) {
             property "sonar.coverage.jacoco.xmlReportPaths", "${rootDir}/build/reports/jacoco/jacocoRootReport/jacocoRootReport.xml"
 
             if (isTravis && isTravisPullRequest) {
-
-                if (hasGithub) {
-                    property "sonar.analysis.mode", "preview"
-                    property "sonar.github.oauth", System.getenv("GITHUBTOKEN")
-                    property "sonar.github.repository", System.getenv("GITHUBREPO")
-                    property "sonar.github.pullRequest", pullRequestId
-                } else {
-                    logger.error("Cannot use sonar github plugin without GITHUBTOKEN (oauth token) and GITHUBREPO (e.g. jbake-org/jbake) environment variables. See https://docs.sonarqube.org/display/PLUG/GitHub+Plugin")
-                }
+                property "sonar.pullrequest.base", "master"
+                property "sonar.pullrequest.branch", "${pullRequestBranch}"
+                property "sonar.pullrequest.key", "${pullRequestId}"
             }
-
         }
     }
 }

--- a/jbake-core/src/main/java/org/jbake/parser/MarkupEngine.java
+++ b/jbake-core/src/main/java/org/jbake/parser/MarkupEngine.java
@@ -129,7 +129,7 @@ public abstract class MarkupEngine implements ParserEngine {
         if (context.getTags() != null) {
             String[] tags = (String[]) context.getTags();
             for (int i = 0; i < tags.length; i++) {
-                tags[i] = sanitizeValue(tags[i]);
+                tags[i] = sanitize(tags[i]);
                 if (context.getConfig().getSanitizeTag()) {
                     tags[i] = tags[i].replace(" ", "-");
                 }
@@ -231,15 +231,15 @@ public abstract class MarkupEngine implements ParserEngine {
     }
 
     private boolean hasHeaderSeparator(String line) {
-        return line.equals(configuration.getHeaderSeparator());
+        return sanitize(line).equals(configuration.getHeaderSeparator());
     }
 
     private boolean isStatusProperty(String line) {
-        return line.startsWith("status=");
+        return sanitize(line).startsWith("status=");
     }
 
     private boolean isTypeProperty(String line) {
-        return line.startsWith("type=");
+        return sanitize(line).startsWith("type=");
     }
 
     /**
@@ -265,8 +265,8 @@ public abstract class MarkupEngine implements ParserEngine {
     }
 
     void storeHeaderValue(String inputKey, String inputValue, Map<String, Object> content) {
-        String key = sanitizeKey(inputKey);
-        String value = sanitizeValue(inputValue);
+        String key = sanitize(inputKey);
+        String value = sanitize(inputValue);
 
         if (key.equalsIgnoreCase(Crawler.Attributes.DATE)) {
             DateFormat df = new SimpleDateFormat(configuration.getDateFormat());
@@ -285,24 +285,19 @@ public abstract class MarkupEngine implements ParserEngine {
         }
     }
 
-    private String sanitizeValue(String part) {
-        return part.trim();
-    }
-
-    private String sanitizeKey(String part) {
-        String key;
+    private String sanitize(String part) {
         if (part.contains(UTF_8_BOM)) {
-            key = part.trim().replace(UTF_8_BOM, "");
+            return part.replace(UTF_8_BOM, "").trim();
         } else {
-            key = part.trim();
+            return part.trim();
         }
-        return key;
     }
 
     private String[] getTags(String tagsPart) {
         String[] tags = tagsPart.split(",");
-        for (int i = 0; i < tags.length; i++)
-            tags[i] = sanitizeValue(tags[i]);
+        for (int i = 0; i < tags.length; i++) {
+            tags[i] = sanitize(tags[i]);
+        }
         return tags;
     }
 

--- a/jbake-core/src/main/java/org/jbake/parser/MarkupEngine.java
+++ b/jbake-core/src/main/java/org/jbake/parser/MarkupEngine.java
@@ -17,6 +17,7 @@ import java.io.InputStream;
 import java.text.DateFormat;
 import java.text.ParseException;
 import java.text.SimpleDateFormat;
+import java.util.Collections;
 import java.util.Date;
 import java.util.List;
 import java.util.Map;
@@ -76,28 +77,17 @@ public abstract class MarkupEngine implements ParserEngine {
      */
     public Map<String, Object> parse(JBakeConfiguration config, File file) {
         this.configuration = config;
-        List<String> fileContents;
-        try (InputStream is = new FileInputStream(file)) {
+        List<String> fileContent = getFileContent(file, config.getRenderEncoding());
 
-            fileContents = IOUtils.readLines(is, config.getRenderEncoding());
-        } catch (IOException e) {
-            LOGGER.error("Error while opening file {}", file, e);
-
+        if (fileContent.isEmpty()) {
             return null;
         }
 
-        boolean hasHeader = hasHeader(fileContents);
-        ParserContext context = new ParserContext(
-                file,
-                fileContents,
-                config,
-                hasHeader
-        );
+        boolean hasHeader = hasHeader(fileContent);
+        ParserContext context = new ParserContext(file, fileContent, config, hasHeader);
 
-        if (hasHeader) {
-            // read header from file
-            processDefaultHeader(context);
-        }
+        // read header from file
+        processDefaultHeader(context);
         // then read engine specific headers
         processHeader(context);
 
@@ -123,6 +113,26 @@ public abstract class MarkupEngine implements ParserEngine {
         // TODO: post parsing plugins to hook in here?
 
         return context.getDocumentModel();
+    }
+
+    private List<String> getFileContent(File file, String encoding) {
+
+        try (InputStream is = new FileInputStream(file)) {
+            LOGGER.debug("read file '{}' with encoding '{}'", file, encoding);
+            List<String> lines = IOUtils.readLines(is, encoding);
+            if (!lines.isEmpty() && isUtf8WithBOM(encoding, lines.get(0))) {
+                LOGGER.warn("remove BOM from file '{}' read with encoding '{}'", file, encoding);
+                lines.set(0, lines.get(0).replace(UTF_8_BOM, ""));
+            }
+            return lines;
+        } catch (IOException e) {
+            LOGGER.error("Error while opening file {}", file, e);
+            return Collections.emptyList();
+        }
+    }
+
+    private boolean isUtf8WithBOM(String encoding, String line) {
+        return encoding.equals("UTF-8") && line.startsWith(UTF_8_BOM);
     }
 
     private void sanitizeTags(ParserContext context) {
@@ -201,10 +211,6 @@ public abstract class MarkupEngine implements ParserEngine {
         return !configuration.getDefaultStatus().isEmpty();
     }
 
-    private boolean hasHeaderSeparatorInContent(List<String> contents) {
-        return contents.indexOf(configuration.getHeaderSeparator()) != -1;
-    }
-
     /**
      * Checks if header separator demarcates end of metadata header
      *
@@ -220,7 +226,7 @@ public abstract class MarkupEngine implements ParserEngine {
 
             for (String line : subContents) {
                 // header should only contain empty lines or lines with '=' in
-                if (!line.contains("=") && !line.isEmpty())  {
+                if (!line.contains("=") && !line.isEmpty()) {
                     return false;
                 }
             }
@@ -248,12 +254,14 @@ public abstract class MarkupEngine implements ParserEngine {
      * @param context the parser context
      */
     private void processDefaultHeader(ParserContext context) {
-        for (String line : context.getFileLines()) {
+        if (context.hasHeader()) {
+            for (String line : context.getFileLines()) {
 
-            if (hasHeaderSeparator(line)) {
-                break;
+                if (hasHeaderSeparator(line)) {
+                    break;
+                }
+                processHeaderLine(line, context.getDocumentModel());
             }
-            processHeaderLine(line, context.getDocumentModel());
         }
     }
 
@@ -286,19 +294,11 @@ public abstract class MarkupEngine implements ParserEngine {
     }
 
     private String sanitize(String part) {
-        if (part.contains(UTF_8_BOM)) {
-            return part.replace(UTF_8_BOM, "").trim();
-        } else {
-            return part.trim();
-        }
+        return part.trim();
     }
 
     private String[] getTags(String tagsPart) {
-        String[] tags = tagsPart.split(",");
-        for (int i = 0; i < tags.length; i++) {
-            tags[i] = sanitize(tags[i]);
-        }
-        return tags;
+        return tagsPart.split(",");
     }
 
     private boolean isJson(String part) {

--- a/jbake-core/src/test/java/org/jbake/app/AsciidocParserTest.java
+++ b/jbake-core/src/test/java/org/jbake/app/AsciidocParserTest.java
@@ -4,8 +4,6 @@ import org.jbake.TestUtils;
 import org.jbake.app.configuration.ConfigUtil;
 import org.jbake.app.configuration.DefaultJBakeConfiguration;
 import org.jbake.app.configuration.JBakeProperty;
-import org.json.simple.JSONArray;
-import org.json.simple.JSONObject;
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Rule;
@@ -14,9 +12,6 @@ import org.junit.rules.TemporaryFolder;
 
 import java.io.File;
 import java.io.PrintWriter;
-import java.util.AbstractMap.SimpleEntry;
-import java.util.Calendar;
-import java.util.Date;
 import java.util.Map;
 
 import static org.assertj.core.api.Assertions.assertThat;
@@ -30,9 +25,16 @@ public class AsciidocParserTest {
     private Parser parser;
     private File rootPath;
 
-    private File validAsciiDocFile;
+    private File asciidocWithSource;
+    private File validAsciidocFile;
+    private File invalidAsciiDocFile;
+    private File validAsciiDocFileWithoutHeader;
+    private File invalidAsciiDocFileWithoutHeader;
+    private File validAsciiDocFileWithHeaderInContent;
+    private File validAsciiDocFileWithoutJBakeMetaData;
 
     private String validHeader = "title=This is a Title = This is a valid Title\nstatus=draft\ntype=post\ndate=2013-09-02\n~~~~~~";
+    private String invalidHeader = "title=This is a Title\n~~~~~~";
 
     @Before
     public void createSampleFile() throws Exception {
@@ -40,8 +42,8 @@ public class AsciidocParserTest {
         config = (DefaultJBakeConfiguration) new ConfigUtil().loadConfig(rootPath);
         parser = new Parser(config);
 
-        validAsciiDocFile = folder.newFile("asciidoc-with-source.ad");
-        PrintWriter out = new PrintWriter(validAsciiDocFile);
+        asciidocWithSource = folder.newFile("asciidoc-with-source.ad");
+        PrintWriter out = new PrintWriter(asciidocWithSource);
         out.println(validHeader);
         out.println("= Hello, AsciiDoc!");
         out.println("Test User <user@test.org>");
@@ -58,6 +60,73 @@ public class AsciidocParserTest {
 
         out.close();
 
+        validAsciidocFile = folder.newFile("valid.ad");
+        out = new PrintWriter(validAsciidocFile);
+        out.println(validHeader);
+        out.println("= Hello, AsciiDoc!");
+        out.println("Test User <user@test.org>");
+        out.println("");
+        out.println("JBake now supports AsciiDoc.");
+        out.close();
+
+        invalidAsciiDocFile = folder.newFile("invalid.ad");
+        out = new PrintWriter(invalidAsciiDocFile);
+        out.println(invalidHeader);
+        out.println("= Hello, AsciiDoc!");
+        out.println("Test User <user@test.org>");
+        out.println("");
+        out.println("JBake now supports AsciiDoc.");
+        out.close();
+
+        validAsciiDocFileWithoutHeader = folder.newFile("validwoheader.ad");
+        out = new PrintWriter(validAsciiDocFileWithoutHeader);
+        out.println("= Hello: AsciiDoc!");
+        out.println("Test User <user@test.org>");
+        out.println("2013-09-02");
+        out.println(":jbake-status: published");
+        out.println(":jbake-type: page");
+        out.println("");
+        out.println("JBake now supports AsciiDoc.");
+        out.close();
+
+        invalidAsciiDocFileWithoutHeader = folder.newFile("invalidwoheader.ad");
+        out = new PrintWriter(invalidAsciiDocFileWithoutHeader);
+        out.println("= Hello, AsciiDoc!");
+        out.println("Test User <user@test.org>");
+        out.println("2013-09-02");
+        out.println(":jbake-status: published");
+        out.println("");
+        out.println("JBake now supports AsciiDoc.");
+        out.close();
+
+        validAsciiDocFileWithHeaderInContent = folder.newFile("validheaderincontent.ad");
+        out = new PrintWriter(validAsciiDocFileWithHeaderInContent);
+        out.println("= Hello, AsciiDoc!");
+        out.println("Test User <user@test.org>");
+        out.println("2013-09-02");
+        out.println(":jbake-status: published");
+        out.println(":jbake-type: page");
+        out.println("");
+        out.println("JBake now supports AsciiDoc.");
+        out.println("");
+        out.println("----");
+        out.println("title=Example Header");
+        out.println("date=2013-02-01");
+        out.println("type=post");
+        out.println("tags=tag1, tag2");
+        out.println("status=published");
+        out.println("~~~~~~");
+        out.println("----");
+        out.close();
+
+        validAsciiDocFileWithoutJBakeMetaData = folder.newFile("validwojbakemetadata.ad");
+        out = new PrintWriter(validAsciiDocFileWithoutJBakeMetaData);
+        out.println("= Hello: AsciiDoc!");
+        out.println("Test User <user@test.org>");
+        out.println("2013-09-02");
+        out.println("");
+        out.println("JBake now supports AsciiDoc documents without JBake meta data.");
+        out.close();
     }
 
 
@@ -65,7 +134,7 @@ public class AsciidocParserTest {
     public void parseAsciidocFileWithPrettifyAttribute() {
 
         config.setProperty(JBakeProperty.ASCIIDOCTOR_ATTRIBUTES,"source-highlighter=prettify");
-        Map<String, Object> map = parser.processFile(validAsciiDocFile);
+        Map<String, Object> map = parser.processFile(asciidocWithSource);
         Assert.assertNotNull(map);
         Assert.assertEquals("draft", map.get("status"));
         Assert.assertEquals("post", map.get("type"));
@@ -74,6 +143,7 @@ public class AsciidocParserTest {
                 .contains("<p>JBake now supports AsciiDoc.</p>")
                 .contains("class=\"prettyprint highlight\"");
 
+        assertThat(map.get("body").toString()).doesNotContain("I Love Jbake");
         System.out.println(map.get("body").toString());
     }
 
@@ -81,7 +151,7 @@ public class AsciidocParserTest {
     public void parseAsciidocFileWithCustomAttribute() {
 
         config.setProperty(JBakeProperty.ASCIIDOCTOR_ATTRIBUTES,"source-highlighter=prettify,testattribute=I Love Jbake");
-        Map<String, Object> map = parser.processFile(validAsciiDocFile);
+        Map<String, Object> map = parser.processFile(asciidocWithSource);
         Assert.assertNotNull(map);
         Assert.assertEquals("draft", map.get("status"));
         Assert.assertEquals("post", map.get("type"));
@@ -92,5 +162,69 @@ public class AsciidocParserTest {
         System.out.println(map.get("body").toString());
     }
 
+    @Test
+    public void parseValidAsciiDocFile() {
+        Map<String, Object> map = parser.processFile(validAsciidocFile);
+        Assert.assertNotNull(map);
+        Assert.assertEquals("draft", map.get("status"));
+        Assert.assertEquals("post", map.get("type"));
+        assertThat(map.get("body").toString())
+            .contains("class=\"paragraph\"")
+            .contains("<p>JBake now supports AsciiDoc.</p>");
+    }
+
+    @Test
+    public void parseInvalidAsciiDocFile() {
+        Map<String, Object> map = parser.processFile(invalidAsciiDocFile);
+        Assert.assertNull(map);
+    }
+
+    @Test
+    public void parseValidAsciiDocFileWithoutHeader() {
+        Map<String, Object> map = parser.processFile(validAsciiDocFileWithoutHeader);
+        Assert.assertNotNull(map);
+        Assert.assertEquals("Hello: AsciiDoc!", map.get("title"));
+        Assert.assertEquals("published", map.get("status"));
+        Assert.assertEquals("page", map.get("type"));
+        assertThat(map.get("body").toString())
+            .contains("class=\"paragraph\"")
+            .contains("<p>JBake now supports AsciiDoc.</p>");
+    }
+
+    @Test
+    public void parseInvalidAsciiDocFileWithoutHeader() {
+        Map<String, Object> map = parser.processFile(invalidAsciiDocFileWithoutHeader);
+        Assert.assertNull(map);
+    }
+
+    @Test
+    public void parseValidAsciiDocFileWithExampleHeaderInContent() {
+        Map<String, Object> map = parser.processFile(validAsciiDocFileWithHeaderInContent);
+        Assert.assertNotNull(map);
+        Assert.assertEquals("published", map.get("status"));
+        Assert.assertEquals("page", map.get("type"));
+        assertThat(map.get("body").toString())
+            .contains("class=\"paragraph\"")
+            .contains("<p>JBake now supports AsciiDoc.</p>")
+            .contains("class=\"listingblock\"")
+            .contains("class=\"content\"")
+            .contains("<pre>")
+            .contains("title=Example Header")
+            .contains("date=2013-02-01")
+            .contains("tags=tag1, tag2");
+    }
+
+    @Test
+    public void parseValidAsciiDocFileWithoutJBakeMetaDataUsingDefaultTypeAndStatus() {
+        config.setDefaultStatus("published");
+        config.setDefaultType("page");
+        Parser parser = new Parser(config);
+        Map<String, Object> map = parser.processFile(validAsciiDocFileWithoutJBakeMetaData);
+        Assert.assertNotNull(map);
+        Assert.assertEquals("published", map.get("status"));
+        Assert.assertEquals("page", map.get("type"));
+        assertThat(map.get("body").toString())
+            .contains("<p>JBake now supports AsciiDoc documents without JBake meta data.</p>");
+    }
 
 }

--- a/jbake-core/src/test/java/org/jbake/app/ParserTest.java
+++ b/jbake-core/src/test/java/org/jbake/app/ParserTest.java
@@ -3,14 +3,13 @@ package org.jbake.app;
 import org.jbake.TestUtils;
 import org.jbake.app.configuration.ConfigUtil;
 import org.jbake.app.configuration.DefaultJBakeConfiguration;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.TemporaryFolder;
-
-import org.json.simple.JSONArray;
-import org.json.simple.JSONObject;
 
 import java.io.File;
 import java.io.PrintWriter;
@@ -32,12 +31,6 @@ public class ParserTest {
 
     private File validHTMLFile;
     private File invalidHTMLFile;
-    private File validAsciiDocFile;
-    private File invalidAsciiDocFile;
-    private File validAsciiDocFileWithoutHeader;
-    private File invalidAsciiDocFileWithoutHeader;
-    private File validAsciiDocFileWithHeaderInContent;
-    private File validAsciiDocFileWithoutJBakeMetaData;
     private File validMarkdownFileWithCustomHeader;
     private File validMarkdownFileWithDefaultStatus;
     private File validMarkdownFileWithDefaultTypeAndStatus;
@@ -47,10 +40,14 @@ public class ParserTest {
     private File validHTMLWithJSONFile;
     private File validAsciiDocWithJSONFile;
     private File validAsciiDocWithADHeaderJSONFile;
+    private File validaAsciidocWithUnsanitizedHeader;
 
     private String validHeader = "title=This is a Title = This is a valid Title\nstatus=draft\ntype=post\ndate=2013-09-02\n~~~~~~";
     private String invalidHeader = "title=This is a Title\n~~~~~~";
     private String sampleJsonData = "{\"numberValue\": 42, \"stringValue\": \"Answer to live, the universe and everything\", \"nullValue\": null, \"arrayValue\": [1, 2], \"objectValue\": {\"val1\": 1, \"val2\": 2}}";
+
+    private String unsanitizedKeys = " title= Title \n status= draft \n   type= post   \ndate=2020-02-30\ncustom=custom without bom's\n~~~~~~";
+
     private String customHeaderSeparator;
 
 
@@ -69,74 +66,6 @@ public class ParserTest {
         invalidHTMLFile = folder.newFile("invalid.html");
         out = new PrintWriter(invalidHTMLFile);
         out.println(invalidHeader);
-        out.close();
-
-        validAsciiDocFile = folder.newFile("valid.ad");
-        out = new PrintWriter(validAsciiDocFile);
-        out.println(validHeader);
-        out.println("= Hello, AsciiDoc!");
-        out.println("Test User <user@test.org>");
-        out.println("");
-        out.println("JBake now supports AsciiDoc.");
-        out.close();
-
-        invalidAsciiDocFile = folder.newFile("invalid.ad");
-        out = new PrintWriter(invalidAsciiDocFile);
-        out.println(invalidHeader);
-        out.println("= Hello, AsciiDoc!");
-        out.println("Test User <user@test.org>");
-        out.println("");
-        out.println("JBake now supports AsciiDoc.");
-        out.close();
-
-        validAsciiDocFileWithoutHeader = folder.newFile("validwoheader.ad");
-        out = new PrintWriter(validAsciiDocFileWithoutHeader);
-        out.println("= Hello: AsciiDoc!");
-        out.println("Test User <user@test.org>");
-        out.println("2013-09-02");
-        out.println(":jbake-status: published");
-        out.println(":jbake-type: page");
-        out.println("");
-        out.println("JBake now supports AsciiDoc.");
-        out.close();
-
-        invalidAsciiDocFileWithoutHeader = folder.newFile("invalidwoheader.ad");
-        out = new PrintWriter(invalidAsciiDocFileWithoutHeader);
-        out.println("= Hello, AsciiDoc!");
-        out.println("Test User <user@test.org>");
-        out.println("2013-09-02");
-        out.println(":jbake-status: published");
-        out.println("");
-        out.println("JBake now supports AsciiDoc.");
-        out.close();
-
-        validAsciiDocFileWithHeaderInContent = folder.newFile("validheaderincontent.ad");
-        out = new PrintWriter(validAsciiDocFileWithHeaderInContent);
-        out.println("= Hello, AsciiDoc!");
-        out.println("Test User <user@test.org>");
-        out.println("2013-09-02");
-        out.println(":jbake-status: published");
-        out.println(":jbake-type: page");
-        out.println("");
-        out.println("JBake now supports AsciiDoc.");
-        out.println("");
-        out.println("----");
-        out.println("title=Example Header");
-        out.println("date=2013-02-01");
-        out.println("type=post");
-        out.println("tags=tag1, tag2");
-        out.println("status=published");
-        out.println("~~~~~~");
-        out.println("----");
-        out.close();
-
-        validAsciiDocFileWithoutJBakeMetaData = folder.newFile("validwojbakemetadata.ad");
-        out = new PrintWriter(validAsciiDocFileWithoutJBakeMetaData);
-        out.println("= Hello: AsciiDoc!");
-        out.println("Test User <user@test.org>");
-        out.println("2013-09-02");
-        out.println("");
-        out.println("JBake now supports AsciiDoc documents without JBake meta data.");
         out.close();
 
         validMarkdownFileWithCustomHeader = folder.newFile("validMdCustomHeader.md");
@@ -257,7 +186,23 @@ public class ParserTest {
         out.println("");
         out.println("JBake now supports AsciiDoc.");
         out.close();
+
+        validaAsciidocWithUnsanitizedHeader = folder.newFile("validAsciidocWithUnsanitizedHeader.adoc");
+        out = new PrintWriter(validaAsciidocWithUnsanitizedHeader, "UTF-8");
+        // Simulating a \uFEFF Byte order Marker in utf-8
+        out.print("\uFEFF");
+        out.println(unsanitizedKeys);
+        out.println("= Hello: AsciiDoc!");
+        out.println("Test User <user@test.org>");
+        out.println("2013-09-02");
+        out.println(":jbake-status: published");
+        out.println(":jbake-type: page");
+        out.print(":jbake-jsondata: ");
+        out.println(sampleJsonData);
+        out.println("");
+        out.println("JBake now supports AsciiDoc.");
         out.close();
+
     }
 
     @Test
@@ -283,78 +228,11 @@ public class ParserTest {
     }
 
     @Test
-    public void parseValidAsciiDocFile() {
-        Map<String, Object> map = parser.processFile(validAsciiDocFile);
-        Assert.assertNotNull(map);
-        Assert.assertEquals("draft", map.get("status"));
-        Assert.assertEquals("post", map.get("type"));
-        assertThat(map.get("body").toString())
-                .contains("class=\"paragraph\"")
-                .contains("<p>JBake now supports AsciiDoc.</p>");
-//        Assert.assertEquals("<div id=\"preamble\">\n<div class=\"sectionbody\">\n<div class=\"paragraph\">\n<p>JBake now supports AsciiDoc.</p>\n</div>\n</div>\n</div>", map.get("body"));
-    }
-
-    @Test
-    public void parseInvalidAsciiDocFile() {
-        Map<String, Object> map = parser.processFile(invalidAsciiDocFile);
-        Assert.assertNull(map);
-    }
-
-    @Test
     public void parseInvalidExtension(){
         Map<String, Object> map = parser.processFile(invalidExtensionFile);
         Assert.assertNull(map);
     }
 
-    @Test
-    public void parseValidAsciiDocFileWithoutHeader() {
-        Map<String, Object> map = parser.processFile(validAsciiDocFileWithoutHeader);
-        Assert.assertNotNull(map);
-        Assert.assertEquals("Hello: AsciiDoc!", map.get("title"));
-        Assert.assertEquals("published", map.get("status"));
-        Assert.assertEquals("page", map.get("type"));
-        assertThat(map.get("body").toString())
-                .contains("class=\"paragraph\"")
-                .contains("<p>JBake now supports AsciiDoc.</p>");
-//        Assert.assertEquals("<div id=\"preamble\">\n<div class=\"sectionbody\">\n<div class=\"paragraph\">\n<p>JBake now supports AsciiDoc.</p>\n</div>\n</div>\n</div>", map.get("body"));
-    }
-
-    @Test
-    public void parseInvalidAsciiDocFileWithoutHeader() {
-        Map<String, Object> map = parser.processFile(invalidAsciiDocFileWithoutHeader);
-        Assert.assertNull(map);
-    }
-
-    @Test
-    public void parseValidAsciiDocFileWithExampleHeaderInContent() {
-        Map<String, Object> map = parser.processFile(validAsciiDocFileWithHeaderInContent);
-        Assert.assertNotNull(map);
-        Assert.assertEquals("published", map.get("status"));
-        Assert.assertEquals("page", map.get("type"));
-        assertThat(map.get("body").toString())
-                .contains("class=\"paragraph\"")
-                .contains("<p>JBake now supports AsciiDoc.</p>")
-                .contains("class=\"listingblock\"")
-                .contains("class=\"content\"")
-                .contains("<pre>")
-                .contains("title=Example Header")
-                .contains("date=2013-02-01")
-                .contains("tags=tag1, tag2");
-//        Assert.assertEquals("<div id=\"preamble\">\n<div class=\"sectionbody\">\n<div class=\"paragraph\">\n<p>JBake now supports AsciiDoc.</p>\n</div>\n<div class=\"listingblock\">\n<div class=\"content\">\n<pre>title=Example Header\ndate=2013-02-01\ntype=post\ntags=tag1, tag2\nstatus=published\n~~~~~~</pre>\n</div>\n</div>\n</div>\n</div>", map.get("body"));
-    }
-
-    @Test
-    public void parseValidAsciiDocFileWithoutJBakeMetaDataUsingDefaultTypeAndStatus() {
-        config.setDefaultStatus("published");
-        config.setDefaultType("page");
-        Parser parser = new Parser(config);
-        Map<String, Object> map = parser.processFile(validAsciiDocFileWithoutJBakeMetaData);
-        Assert.assertNotNull(map);
-        Assert.assertEquals("published", map.get("status"));
-        Assert.assertEquals("page", map.get("type"));
-        assertThat(map.get("body").toString())
-                .contains("<p>JBake now supports AsciiDoc documents without JBake meta data.</p>");
-    }
 
     @Test
     public void parseMarkdownFileWithCustomHeaderSeparator() {
@@ -403,6 +281,16 @@ public class ParserTest {
     public void parseInvalidMarkdownFile() {
         Map<String, Object> map = parser.processFile(invalidMDFile);
         Assert.assertNull(map);
+    }
+
+    @Test
+    public void sanitizeKeysAndValues() {
+        Map<String, Object> map = parser.processFile(validaAsciidocWithUnsanitizedHeader);
+
+        assertThat(map.get("status")).isEqualTo("draft");
+        assertThat(map.get("title")).isEqualTo("Title");
+        assertThat(map.get("type")).isEqualTo("post");
+        assertThat(map.get("custom")).isEqualTo("custom without bom's");
     }
 
     @Test

--- a/jbake-core/src/test/java/org/jbake/app/ParserTest.java
+++ b/jbake-core/src/test/java/org/jbake/app/ParserTest.java
@@ -3,6 +3,7 @@ package org.jbake.app;
 import org.jbake.TestUtils;
 import org.jbake.app.configuration.ConfigUtil;
 import org.jbake.app.configuration.DefaultJBakeConfiguration;
+import org.jbake.app.configuration.JBakeProperty;
 import org.json.simple.JSONArray;
 import org.json.simple.JSONObject;
 import org.junit.Assert;
@@ -14,6 +15,7 @@ import org.junit.rules.TemporaryFolder;
 import java.io.File;
 import java.io.PrintWriter;
 import java.util.AbstractMap.SimpleEntry;
+import java.util.Arrays;
 import java.util.Calendar;
 import java.util.Date;
 import java.util.Map;
@@ -46,7 +48,7 @@ public class ParserTest {
     private String invalidHeader = "title=This is a Title\n~~~~~~";
     private String sampleJsonData = "{\"numberValue\": 42, \"stringValue\": \"Answer to live, the universe and everything\", \"nullValue\": null, \"arrayValue\": [1, 2], \"objectValue\": {\"val1\": 1, \"val2\": 2}}";
 
-    private String unsanitizedKeys = " title= Title \n status= draft \n   type= post   \ndate=2020-02-30\ncustom=custom without bom's\n~~~~~~";
+    private String unsanitizedKeys = " title= Title \n status= draft \n   type= post   \ndate=2020-02-30\ncustom=custom without bom's\ntags= jbake, java    , tag with space   \n~~~~~~";
 
     private String customHeaderSeparator;
 
@@ -291,7 +293,17 @@ public class ParserTest {
         assertThat(map.get("title")).isEqualTo("Title");
         assertThat(map.get("type")).isEqualTo("post");
         assertThat(map.get("custom")).isEqualTo("custom without bom's");
+        assertThat(map.get("tags")).isEqualTo(Arrays.asList("jbake", "java", "tag with space").toArray());
     }
+
+    @Test
+    public void sanitizeTags() {
+        config.setProperty(JBakeProperty.TAG_SANITIZE, true);
+        Map<String, Object> map = parser.processFile(validaAsciidocWithUnsanitizedHeader);
+
+        assertThat(map.get("tags")).isEqualTo(Arrays.asList("jbake", "java", "tag-with-space").toArray());
+    }
+
 
     @Test
     public void parseValidHTMLWithJSONFile() {

